analysis/dataflow/state.go

```diff
@@ -82,90 +56,42 @@ type AnalyzerState struct {
 	// a map
 	BoundingInfo BoundingMap
 
-	// Report contains the accumulated report information
-	Report *config.ReportInfo
-
 	reachableFunctions map[*ssa.Function]bool
-
-	// a callgraph computed using the cha analysis. Useful to boostrap the reachable functions
-	chaCallgraph      *callgraph.Graph
-	isReachabilityCha bool
-
-	// Stored errors
-	numAlarms  atomic.Int32
-	errors     map[string][]error
-	errorMutex sync.Mutex
 }
 
-// NewInitializedAnalyzerState runs NewAnalyzerState, and any additional steps that are commonly used in analyses.
+// NewState generates a State from a PointerState
 // This consists in:
-//   - building the ssa program
-//   - running pointer analysis
 //   - computing a map from interface types to the implementations of their methods
 //   - scanning the usage of globals in the program
 //   - linking aliases of bound variables to the closure that binds them
-func NewInitializedAnalyzerState(program *ssa.Program, pkgs []*packages.Package,
-	logger *config.LogGroup, config *config.Config) (*AnalyzerState, error) {
-	program.Build()
-	state, err := NewAnalyzerState(program, pkgs, logger, config, []func(*AnalyzerState){
-		func(s *AnalyzerState) { s.PopulateImplementations() },
-		func(s *AnalyzerState) { s.PopulatePointersVerbose(summaries.IsUserDefinedFunction) },
-		func(s *AnalyzerState) { s.PopulateGlobalsVerbose() },
+//
+// The State returned *does not* have dataflow information computed yet.
+func NewState(ps *ptr.State) result.Result[State] {
+	state, err := newFlowState(*ps, []func(*State){
```

Comment:
nit: Is there ever an instance where you don't populate the implementations, globals, and bounding information? I think this function can be simplified a lot by calling these functions explicitly instead of passing them in

Reply:
There's no such instance, but this approach makes it easier to "declare" the parallel steps that need to be run in the high-level `NewState` function and then let `newFlowState` apply them in parallel. I don't think we'd gain a lot in readability by making the function calls explicit with separate go calls and increment/decrements of the waitgroup inside the `newFlowState` function. This also makes it easy to add more steps.